packages/web-worker/src/endpoint.ts

```diff
@@ -146,6 +155,13 @@ export function createEndpoint<T>(
       makeCallable(port, () => newValue);
       functionStore.set(newValue, [id, port]);
     },
+    terminateEndpoint() {
+      if (messageEndpoint.terminate) {
+        [functionStore, functionProxies, activeApi].forEach(map => map.clear());
+        messageEndpoint.terminate();
+        terminated = true;
```

Comment:
We should definitely be doing this and the clearing of all the stores whether the `messageEndoint` has `terminate` or not.

Reply:
Also, I still think the main thread should be notified when `terminate` is called in the worker. At the very least, it lets the main thread release all of its function store/ proxies/ activeApi/ etc

Reply:
Sounds fair. Important to note that calling `terminate` from worker doesn't truely terminate the worker as [mentioned earlier](https://github.com/Shopify/quilt/pull/1111#discussion_r335038322). It just clears its maps and sets `terminated` to true, preventing subsequent worker calls. 

Updated PR with tests.